### PR TITLE
MemoryAllocation & other assorted improvements.

### PR DESCRIPTION
**Pull request by redknightlois**
- Improvements on Hashing.Combine using a ROL assembler instruction instead.
- Improved memory allocation pooling safe structures at the LowLevelTransaction level.
- Improved PageFromScratchBuffer lookups adding a missing comparer.

### CODE REVIEW

File: src/Voron/Impl/LowLevelTransaction.cs [hunk 1]
```diff
@@ -21,27 +21,40 @@ namespace Voron.Impl
...
+        public Tree RootObjects => _root;        
 
-        public bool FlushedToJournal;
-        public Tree RootObjects => _root;
+        public bool FlushedToJournal;        
 
         private readonly WriteAheadJournal _journal;
+        internal readonly List<JournalSnapshot> JournalSnapshots = new List<JournalSnapshot>();
 
+        private static ObjectPool<Dictionary<long, PageFromScratchBuffer>, DictionaryResetBehavior<long, PageFromScratchBuffer>> ScratchPagesTablePool = new ObjectPool<Dictionary<long, PageFromScratchBuffer>, DictionaryResetBehavior<long, PageFromScratchBuffer>>(() => new Dictionary<long, PageFromScratchBuffer>(NumericEqualityComparer.Instance), 30);
```
Comment (ayende) [inline src/Voron/Impl/LowLevelTransaction.cs:39]: Why is this static? Isn't it better to do it at the level of the Environment?

My concern is that it is quite easy to have more than 30 storage environments, and then we'll have a lot of churn.

**Comment by redknightlois** [comment]
Added extra improvements.

File: src/Voron/Impl/LowLevelTransaction.cs (hunk 1 above)
Comment (redknightlois) [reply src/Voron/Impl/LowLevelTransaction.cs:39]: Thats actually a good idea. In that way each pool will behave closely to its typical use. And we can also react to low memory conditions

Comment (ayende) [reply src/Voron/Impl/LowLevelTransaction.cs:39]: Okay, I got something imlemented

**Pull request merged**